powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationServiceBehavior.java

```diff
@@ -139,9 +140,19 @@ public OperationDetailResponse createOperation(OperationCreateRequest request) t
             throw localizationProvider.buildExceptionForCode(ServiceError.UNABLE_TO_GENERATE_TOKEN);
         }
 
-        // Get operation expiration date
-        final long expiration = templateEntity.getExpiration() * 1000L;
-        final Date timestampExpiration = new Date(currentTimestamp.getTime() + expiration);
+        // Resolve the operation expiration date
+        final Date timestampExpires;
+        if (timestampExpiresRequest != null) {
+            timestampExpires = timestampExpiresRequest;
+        } else {
+            final long expiration = templateEntity.getExpiration() * 1000L;
+            timestampExpires = new Date(currentTimestamp.getTime() + expiration);
+        }
+
+        if (timestampExpires.before(new Date())) {
```

Comment:
Use `currentTimestamp` value instead of `new Date()`.

Reply:
Also, the validation can probably happen early on in the request.

Reply:
Good points.